### ISSUE TITLE
Acessing a non existing instance attribute create a trait instance in the class

### ISSUE THREAD

**Issue by sapetnioc** [enthought/traits#358]
The following code creates two instances of `HasTrait` in `an_instance` and `another_instance` (problem is the same with a derived class in a real case). Then, accessing `an_instance.i_do_not_exists` creates a new trait in the class that is accessible via `another_instance`. Any attempt to read a non existing instance attribute creates a new item in the class `__dict__`.

Moreover, a subsequent use of `add_trait` to create a trait named `i_do_not_exists` in any instance does not add it to the list of traits returned by `traits()`. 

```
from traits.api import HasTraits, String

an_instance = HasTraits()
another_instance = HasTraits()

try:
      an_instance.i_do_not_exist
except AttributeError:
    pass

print another_instance.trait('i_do_not_exist')
# -> <traits.traits.CTrait object at 0x7f65937315d0>

another_instance.add_trait('i_do_not_exist', String)
print another_instance.traits().keys()
# -> ['trait_added', 'trait_modified']
```

**Comment by corranwebster** [comment]
`HasTraits` is intended to be an abstract base class isn't intended to be directly instantiated; but I assume that you are seeing something similar in a concrete subclass of `HasTraits`?

In any case, I don't see the behaviour you describe for `add_trait`:
```
In [158]: from traits.api import HasTraits, String
     ...: 
     ...: an_instance = HasTraits()
     ...: another_instance = HasTraits()
     ...: 
     ...: try:
     ...:       an_instance.i_do_not_exist
     ...: except AttributeError:
     ...:     pass
     ...: 
     ...: print another_instance.trait('i_do_not_exist')
     ...: # -> <traits.traits.CTrait object at 0x7f65937315d0>
     ...: 
     ...: another_instance.add_trait('i_do_not_exist', String)
     ...: print another_instance.traits().keys()
     ...: # -> ['trait_added', 'trait_modified']
<traits.traits.CTrait object at 0x11b9701b0>
['trait_added', 'i_do_not_exist', 'trait_modified']
```

Can you give more information about the version of traits that you are using or other information which may be relevant to reproduce the issue?

**Comment by sapetnioc** [comment]
I used HasTraits just to make the example code shorter but the behavior is the same with subclasses.

I tried the sample code Traits version of Ubuntu 16.04 (traits version 4.5.0 according to `pip freeze`) and also traits 4.6.0 in a virtualenv (on the same system). A colleague told me that the behavior is the same on Ubuntu 16.10. 

Just to be sure my own environment is not involved, I just tested the code on a minimum Ubuntu 16.04 system via Docker and had the same behavior.
```

```

**Comment by corranwebster** [comment]
Hmmm... still can't replicate.  Originally I was using EPD's version of Traits 4.6 on OS X, but in a clean environment on Ubuntu 16.04 (ie. I had to apt-get install pip and python) and a pip installed version of Traits 4.6 I am still not seeing the issue.  I am running these tests under Python 2.7 (and I assume that you are doing the same, based on the `print` statements).
```
>>> from traits.api import HasTraits, String
>>> 
>>> an_instance = HasTraits()
>>> another_instance = HasTraits()
>>> 
>>> try:
...       an_instance.i_do_not_exist
... except AttributeError:
...     pass
... 
>>> print another_instance.trait('i_do_not_exist')
<traits.traits.CTrait object at 0x7feb736f1d60>
>>> # -> <traits.traits.CTrait object at 0x7f65937315d0>
... 
>>> another_instance.add_trait('i_do_not_exist', String)
>>> print another_instance.traits().keys()
['trait_added', 'i_do_not_exist', 'trait_modified']
```

Is there any other information you can give that might help replicate & diagnose? At this point the only thing I can think of is perhaps different versions of gcc?

**Comment by corranwebster** [comment]
Looking more closely, I am also not seeing the `'i_do_not_exist'` trait being added to the `__dict__` of `an_instance`:
```
In [170]: an_instance.__dict__
Out[170]: {}
```
Nor is it added to the traits of `an_instance`:
```
In [171]: an_instance.traits()
Out[171]: 
{'trait_added': <traits.traits.CTrait at 0x11b970050>,
 'trait_modified': <traits.traits.CTrait at 0x11b965f70>}
```

**Comment by denisri** [comment]
Using either Ubuntu 16.04 or Ubuntu 16.10, python 2.7 and Traits from the distribution (apt-get install python-traits), and running in IPython (would it matter?), I have the same problem. For Ubuntu 16.04 it is traits version 4.5.0.

**Comment by corranwebster** [comment]
Ah. I _can_ replicate on the version I get from doing `apt-get install python-traits` and with a pip-installed Traits 4.5 on Ubuntu.

@denisri or @sapetnioc can either of you confirm that you can reproduce using Traits 4.6, and if so under what conditions?

**Comment by sapetnioc** [comment]
You are right, I made a mistake when I told that the problem is exactly the same with 4.6, sorry. Indeed, the class is modified when accessing the non existent attribute but `add_trait` works correctly on `another_trait` (I.e. `i_do_not_exist` is in the list returned by `traits()`). However, I believe that adding a trait to the class when accessing an instance attribute is still a bug for two reasons : 
- it is a possible source of memory leak
- using `trait[name)` to test the existence of a trait with a given name can lead to a false positive
